cmd/api/main.go

```diff
@@ -117,6 +120,20 @@ func main() {
 		log.Fatalf("Failed to connect to database: %v", err)
 	}
 
+	sqlDB, err := db.DB()
+
+	// Set DB connection limits
+	maxIdle := serverConfig.DB_MAX_IDLE_CONNECTIONS
+	maxOpen := serverConfig.DB_MAX_OPEN_CONNECTIONS
+	if maxOpen < 1 {
+		maxOpen = defaultDatabaseConnections
```

Comment:
Overloading the database with more and more connections is bad for DB performance, you are better off having some goroutines wait for a connection than just send everything at once.

In additional postgres always has a limit of how many connections it has been configured to accept, even if the default here is unlimited at some point you start getting errors.

Reply:
I meant not putting limits if not specified in the config.


Is 10 sufficient? I am not sure. 
https://www.postgresql.org/docs/current/runtime-config-connection.html#GUC-MAX-CONNECTIONS

Documentation shows that we have 100 by default in PG. So should we put this a bit higher?

Reply:
performance testing last nigh reference our new internal slack channel proved otherwise @khrm wrt 100 being a default in PG

IMO 10 is a safer default, given the need to support developer mode out of the gate 

we can tune it higher using @stuartwdouglas 's change here in specific production envs if performance testing dictates such a change would help

if possible @sayan-biswas @khrm @avinal @enarha let's prioritize review / iterate on any further review / merging

I can talk about his further in the WG call Thursday if need be if this is still pending by then.

Thanks.

Reply:
Did we get the performance boost from IdleMax increase of maxOpen increase? Didn't we try the whole PR?

Reply:
We should try maxIdle vs maxIdleTime vs maxOpen + maxIdle.

```
Keep in mind that setting a limit makes database usage similar to acquiring a lock or semaphore, with the result that your application can deadlock waiting for a new database connection.
```

Reply:
Fixed sized thread pools are always the fastest. At the end of the day the database will have a point where performance starts to degrade with more connections rather than increase, and it becomes better to queue requests rather than creating more connections (in addition databases will always have some kind of connection limit). 

Variable sized thread pools means you start creating connections (which is very slow, possibly as mush as 100ms) right when both your app and the DB is under the most load. If you know you might need up to 'n' connections then you are way better off just starting 'n' connections, especially if you are the only application using the database.

Also at the moment in production the API server seems to top out at about 5-10 req/sec while the DB part of that is milliseconds, so 10 will be fine to start with. With the new index the slow 20m+ queries we were seeing are now around the 50ms mark (with most queries being way faster, e.g. GetRecord is around 1ms), so even if all you are running is the really slow queries 10 connections is still enough to handle 200req/sec, which is an order of magniture  higher than what the API server is currently able to handle.

Reply:
> We should try maxIdle vs maxIdleTime vs maxOpen + maxIdle.
> 
> ```
> Keep in mind that setting a limit makes database usage similar to acquiring a lock or semaphore, with the result that your application can deadlock waiting for a new database connection.
> ```

It won't deadlock, its a queue, unless you start trying to acquire multiple connections from the same goroutine, which we are not doing (e.g. if you have a limit of 10, and a goroutine tries to acquire 11 all at once). Before Gabes patch that raised to kube client QPS there is already a queue that limits us to 5 queries per second, so this is not really a concern.

Reply:
If we assume RedHat Pipeline Service, it might make sense. But when DB is local instead of managed, is a new connection still costly that costly? That's what we shipped by default in the product.

It's not deadlock in the sense of not being able to work. It means that a new request goroutine would be blocked on acquiring connection. 

I am not saying that both maxOpen and IdleConn won't help but we shouldn't change the default ourselves assuming Pipelines service load.

Reply:
```
Fixed sized thread pools are always the fastest. At the end of the day the database will have a point where performance starts to degrade with more connections rather than increase, and it becomes better to queue requests rather than creating more connections (in addition databases will always have some kind of connection limit).

```
I agree with this but we shouldn't change the default of the library. Instead, we should allow these settings to be configurable.

Reply:
@khrm the 10 setting that @stuartwdouglas has here as the default is not a setting precisely tuned to our internal environment, the "pipeline service" env.  My recent testing of this change with this value was in a personal test cluster, development mode if you will, where I ran first 1 query, and then 10 more or less concurrent queries against first ***1*** row, and then 10 rows.  We saw improvement with both 1 query and 10 concurrent query tests.

Guaranteed we'll change this number to something greater than 10 in our internal environment.  10 is **NOT** the internal environment or "pipeline service" setting.

@stuartwdouglas and I saw anywhere from 5x improvement (500 ms to 100 ms) to 250x improvement (500 ms to 2 ms) in my development environment.

But consider, not our internal deployment of results, but a new user trying out results for the first time.  The out of the box experience if you will.  They will try simple tests like what I described above.  In their "development environments".  If they see slow response times out of the box, they could be turned off, unmotivated, from using results.  Thinking it is naively implemented.

Some minimal connection related pooling improves this out of the box experience, even in development / personal clusters, but is not so onerous that it cannot be run in development / personal clusters.

I know the WG call is coming in an hour or two as I type.  We can continue the discussion there of course.  But if you see this in the interim, certainly let's continue discussing here.

Reply:
@khrm and I talked during the WG call and the outcome was I convinced him per the reasoning in my last comment.  

However, he would like to try and test out this change in his developer env as part of signing off on this PR.